runtime/src/snapshot_utils.rs

```diff
@@ -536,18 +679,121 @@ where
     Ok(ret)
 }
 
+fn deserialize_snapshot_data_files_capped<T: Sized>(
+    snapshot_root_paths: &SnapshotRootPaths,
+    maximum_file_size: u64,
+    deserializer: impl FnOnce(&mut SnapshotStreams<File>) -> Result<T>,
+) -> Result<T> {
+    let (full_snapshot_file_size, mut full_snapshot_data_file_stream) =
+        prepare_deserialize_snapshot_data_file(
+            &snapshot_root_paths.full_snapshot_root_file_path,
+            maximum_file_size,
+        )?;
+
+    let (incremental_snapshot_file_size, mut incremental_snapshot_data_file_stream) =
+        if let Some(ref incremental_snapshot_root_file_path) =
+            snapshot_root_paths.incremental_snapshot_root_file_path
+        {
+            let (incremental_snapshot_file_size, incremental_snapshot_data_file_stream) =
+                prepare_deserialize_snapshot_data_file(
+                    incremental_snapshot_root_file_path,
+                    maximum_file_size,
+                )?;
+            (
+                Some(incremental_snapshot_file_size),
+                Some(incremental_snapshot_data_file_stream),
+            )
+        } else {
+            (None, None)
+        };
+
+    let mut snapshot_streams = SnapshotStreams {
+        full_snapshot_stream: &mut full_snapshot_data_file_stream,
+        incremental_snapshot_stream: incremental_snapshot_data_file_stream.as_mut(),
+    };
+    let ret = deserializer(&mut snapshot_streams)?;
+
+    check_deserialize_file_consumed(
+        full_snapshot_file_size,
+        &snapshot_root_paths.full_snapshot_root_file_path,
+        &mut full_snapshot_data_file_stream,
+    )?;
+
+    if let Some(ref incremental_snapshot_root_file_path) =
+        snapshot_root_paths.incremental_snapshot_root_file_path
+    {
+        check_deserialize_file_consumed(
+            incremental_snapshot_file_size.unwrap(),
+            incremental_snapshot_root_file_path,
+            incremental_snapshot_data_file_stream.as_mut().unwrap(),
+        )?;
+    }
+
+    Ok(ret)
+}
+
+/// Before running the deserializer function, perform common operations on the snapshot archive
+/// files, such as checking the file size and opening the file into a stream.
+fn prepare_deserialize_snapshot_data_file<P>(
```

Comment:
nits: naming: `create_snapshot_data_file_stream`? `prepare_` is a bit too general. I know you picked `prepare_` because of the extra checks to encompass these functionalities. but as par defensive programming, doing routine checks are obvious unless we're too lazy. xD

Reply:
Goo points. Done!